Adding UA stylesheet for counter-styles (counterStyles.css)

vitorroriz opened the pull request:
#### b73a449ec5ef5e240f89806e332486bfc79b1cac
<pre>
Adding UA stylesheet for counter-styles (counterStyles.css)
<a href="https://bugs.webkit.org/show_bug.cgi?id=252861">https://bugs.webkit.org/show_bug.cgi?id=252861</a>
rdar://105853064

Adding a new UA stylesheet for storing the rules
for the normative simple predefined counter styles
as specified at <a href="https://www.w3.org/TR/css-counter-styles-3/#predefined-counters">https://www.w3.org/TR/css-counter-styles-3/#predefined-counters</a>

Reviewed by Tim Nguyen.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::addUserAgentCounterStyle):
(WebCore::CSSCounterStyleRegistry::userAgentCounterStyles):
* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/counterStyles.css: Added.
(@counter-style decimal):
(@counter-style decimal-leading-zero):
(@counter-style arabic-indic):
(@counter-style armenian):
(@counter-style upper-armenian):
(@counter-style lower-armenian):
(@counter-style bengali):
(@counter-style cambodian):
(@counter-style khmer):
(@counter-style cjk-decimal):
(@counter-style devanagari):
(@counter-style georgian):
(@counter-style gujarati):
(@counter-style gurmukhi):
(@counter-style hebrew):
(@counter-style kannada):
(@counter-style lao):
(@counter-style malayalam):
(@counter-style mongolian):
(@counter-style myanmar):
(@counter-style oriya):
(@counter-style persian):
(@counter-style lower-roman):
(@counter-style upper-roman):
(@counter-style tamil):
(@counter-style telugu):
(@counter-style thai):
(@counter-style tibetan):
(@counter-style lower-alpha):
(@counter-style lower-latin):
(@counter-style upper-alpha):
(@counter-style upper-latin):
(@counter-style lower-greek):
(@counter-style hiragana):
(@counter-style hiragana-iroha):
(@counter-style katakana):
(@counter-style katakana-iroha):
(@counter-style disc):
(@counter-style circle):
(@counter-style square):
(@counter-style disclosure-open):
(@counter-style disclosure-closed):
(@counter-style cjk-earthly-branch):
(@counter-style cjk-heavenly-stem):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::addToCounterStyleRegistry):
(WebCore::Style::UserAgentStyle::ensureDefaultStyleSheetsForElement):
* Source/WebCore/style/UserAgentStyle.h:

Canonical link: <a href="https://commits.webkit.org/260795@main">https://commits.webkit.org/260795@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a2c7b620936df54cedc99c1bb7e0068f596fc05

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109460 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18584 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42222 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/958 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118624 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113345 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20050 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9781 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101736 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115216 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98161 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43137 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29814 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84898 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11328 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31157 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11991 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8092 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17357 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50761 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7478 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13723 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/style/UserAgentStyle.cpp:
```diff
@@ -119,6 +122,14 @@ static StyleSheetContents* parseUASheet(const String& str)
+    for (auto& rule : sheet.childRules()) {
+        if (!is<StyleRuleCounterStyle>(*rule))
+            continue;
+        CSSCounterStyleRegistry::addUserAgentCounterStyle(downcast<StyleRuleCounterStyle>(*rule).descriptors());
```
```suggestion
        if (auto* counterStyleRule = dynamicDowncast<StyleRuleCounterStyle>(rule))
            CSSCounterStyleRegistry::addUserAgentCounterStyle(counterStyleRule->descriptors());
```

nt1m commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -137,7 +142,7 @@ RefPtr<CSSCounterStyle> CSSCounterStyleRegistry::resolvedCounterStyle(const Atom
+    // counter-style should get pre-defined counters from UA stylesheet, but we are inserting decimal manually here because it is the last fallback.
     auto initialStyles = []() {
         CounterStyleMap map;
         map.add("decimal"_s, CSSCounterStyle::createCounterStyleDecimal());
```
I'd prefer that we remove this in favor of the new mechanism, and that we assert that the decimal counter style exists in the places we need it.

nt1m commented on Source/WebCore/css/CSSCounterStyleRegistry.cpp:
```diff
@@ -137,13 +142,6 @@ RefPtr<CSSCounterStyle> CSSCounterStyleRegistry::resolvedCounterStyle(const Atom
-    // FIXME: counter-style should get pre-defined counters from UA stylesheet rdar://103021161.
-    auto initialStyles = []() {
-        CounterStyleMap map;
-        map.add("decimal"_s, CSSCounterStyle::createCounterStyleDecimal());
```
Can you delete this `CSSCounterStyle::createCounterStyleDecimal()` function? I assume it's unused now

nt1m commented on Source/WebCore/css/counterStyles.css:
```diff
@@ -0,0 +1,349 @@
+    suffix: ', ';
+    negative: "\B9C8\C774\B108\C2A4  ";
+    /* 마이너스 (followed by a space) */
+  }
```
ditto

nt1m commented on Source/WebCore/css/counterStyles.css:
```diff
@@ -0,0 +1,349 @@
+    /* 마이너스 (followed by a space) */
+  }
+
+  @counter-style korean-hanja-formal {
```
ditto:
```suggestion
}

@counter-style korean-hanja-formal {
```

nt1m commented on Source/WebCore/css/counterStyles.css:
```diff
@@ -0,0 +1,349 @@
+    /* 마이너스 (followed by a space) */
+  }
+
+  @counter-style korean-hanja-informal {
```
nit: weird indent

```suggestion
}

@counter-style korean-hanja-informal {
```

vitorroriz committed:
Adding UA stylesheet for counter-styles (counterStyles.css)
https://bugs.webkit.org/show_bug.cgi?id=252861
rdar://105853064

Adding a new UA stylesheet for storing the rules
for the normative simple predefined counter styles
as specified at https://www.w3.org/TR/css-counter-styles-3/#predefined-counters

Reviewed by Tim Nguyen.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::addUserAgentCounterStyle):
(WebCore::CSSCounterStyleRegistry::userAgentCounterStyles):
* Source/WebCore/css/CSSCounterStyleRegistry.h:
* Source/WebCore/css/counterStyles.css: Added.
(@counter-style decimal):
(@counter-style decimal-leading-zero):
(@counter-style arabic-indic):
(@counter-style armenian):
(@counter-style upper-armenian):
(@counter-style lower-armenian):
(@counter-style bengali):
(@counter-style cambodian):
(@counter-style khmer):
(@counter-style cjk-decimal):
(@counter-style devanagari):
(@counter-style georgian):
(@counter-style gujarati):
(@counter-style gurmukhi):
(@counter-style hebrew):
(@counter-style kannada):
(@counter-style lao):
(@counter-style malayalam):
(@counter-style mongolian):
(@counter-style myanmar):
(@counter-style oriya):
(@counter-style persian):
(@counter-style lower-roman):
(@counter-style upper-roman):
(@counter-style tamil):
(@counter-style telugu):
(@counter-style thai):
(@counter-style tibetan):
(@counter-style lower-alpha):
(@counter-style lower-latin):
(@counter-style upper-alpha):
(@counter-style upper-latin):
(@counter-style lower-greek):
(@counter-style hiragana):
(@counter-style hiragana-iroha):
(@counter-style katakana):
(@counter-style katakana-iroha):
(@counter-style disc):
(@counter-style circle):
(@counter-style square):
(@counter-style disclosure-open):
(@counter-style disclosure-closed):
(@counter-style cjk-earthly-branch):
(@counter-style cjk-heavenly-stem):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::CSSParserContext::CSSParserContext):
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::addToCounterStyleRegistry):
(WebCore::Style::UserAgentStyle::ensureDefaultStyleSheetsForElement):
* Source/WebCore/style/UserAgentStyle.h:

Canonical link: https://commits.webkit.org/260795@main
